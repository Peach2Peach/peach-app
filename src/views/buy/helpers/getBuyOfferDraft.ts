type GetBuyOfferDraftParams = {
  minBuyAmount: number
  maxBuyAmount: number
  meansOfPayment?: MeansOfPayment
}
export const getBuyOfferDraft = ({
  minBuyAmount,
  maxBuyAmount,
  meansOfPayment = {},
}: GetBuyOfferDraftParams): BuyOfferDraft => ({
  type: 'bid',
  creationDate: new Date(),
  lastModified: new Date(),
  amount: [minBuyAmount, maxBuyAmount],
  meansOfPayment,
  paymentData: {},
  releaseAddress: '',
  originalPaymentData: [],
})
