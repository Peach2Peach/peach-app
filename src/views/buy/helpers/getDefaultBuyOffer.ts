type DefaultOfferProps = {
  minBuyAmount: number
  maxBuyAmount: number
  meansOfPayment?: MeansOfPayment
  kyc?: boolean
}
export const getDefaultBuyOffer = ({
  minBuyAmount,
  maxBuyAmount,
  meansOfPayment = {},
  kyc = false,
}: DefaultOfferProps): BuyOfferDraft => ({
  type: 'bid',
  creationDate: new Date(),
  lastModified: new Date(),
  amount: [minBuyAmount, maxBuyAmount],
  meansOfPayment,
  paymentData: {},
  releaseAddress: '',
  originalPaymentData: [],
  kyc,
})
