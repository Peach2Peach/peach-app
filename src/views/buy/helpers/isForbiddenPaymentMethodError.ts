import { isPaymentMethod } from '../../../utils/validation/isPaymentMethod'

export const isForbiddenPaymentMethodError = (
  errorMessage: string | null,
  errorDetails: unknown,
): errorDetails is PaymentMethod[] =>
  errorMessage === 'FORBIDDEN' && Array.isArray(errorDetails) && errorDetails.every(isPaymentMethod)
