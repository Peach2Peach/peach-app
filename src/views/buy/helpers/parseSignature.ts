const signaturePattern = /[a-zA-Z0-9/=+]{88}/u

export const parseSignature = (signature: string) => signature.match(signaturePattern)?.pop() || signature
