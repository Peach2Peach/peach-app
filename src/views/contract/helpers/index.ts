export { isPSBTForBatch } from "./isPSBTForBatch";
export { shouldShowTradeStatusInfo } from "./shouldShowTradeStatusInfo";
export { verifyReleasePSBT as verifyPSBT } from "./verifyReleasePSBT";
