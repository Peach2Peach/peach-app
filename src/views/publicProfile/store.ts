import create from 'zustand'

type ProfileStore = {
  currentUserPubkey: string
  setCurrentUserPubkey: (currentUserPubkey: string) => void
}

export const useProfileStore = create<ProfileStore>()((set) => ({
  currentUserPubkey: '',
  setCurrentUserPubkey: (currentUserPubkey: string) => set({ currentUserPubkey }),
}))
