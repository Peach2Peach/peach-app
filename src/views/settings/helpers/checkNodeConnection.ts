import { Blockchain } from 'bdk-rn'
import { BlockChainNames, BlockchainElectrumConfig, BlockchainEsploraConfig } from 'bdk-rn/lib/lib/enums'
import { info } from '../../../utils/log'
import { parseError } from '../../../utils/result'
import { addProtocol } from '../../../utils/web'

const checkElectrumConnection = async (address: string, ssl: boolean) => {
  const config: BlockchainElectrumConfig = {
    url: addProtocol(address, ssl ? 'ssl' : 'tcp'),
    sock5: null,
    retry: 1,
    timeout: 5,
    stopGap: 1,
    validateDomain: false,
  }

  try {
    info('Checking electrum connection...')
    const blockchain = await new Blockchain().create(config, BlockChainNames.Electrum)
    await blockchain.getBlockHash()
    return { result: BlockChainNames.Electrum }
  } catch (e) {
    info('electrum connection failed')
    return { error: parseError(e) }
  }
}
const checkEsploraConnection = async (address: string, ssl: boolean) => {
  const config: BlockchainEsploraConfig = {
    baseUrl: addProtocol(address, ssl ? 'https' : 'http'),
    proxy: null,
    concurrency: 1,
    timeout: 5,
    stopGap: 1,
  }

  try {
    info('Checking esplora connection...')
    const blockchain = await new Blockchain().create(config, BlockChainNames.Esplora)
    await blockchain.getBlockHash()
    return { result: BlockChainNames.Esplora }
  } catch (e) {
    info('esplora connection failed')
    return { error: parseError(e) }
  }
}
const connectionChecks = [checkElectrumConnection, checkEsploraConnection]
export const checkNodeConnection = async (address: string, ssl = false) => {
  const errors: unknown[] = []
  for (const connectionCheck of connectionChecks) {
    // eslint-disable-next-line no-await-in-loop
    const { result, error } = await connectionCheck(address, ssl)
    if (result) return { result }
    errors.push(error)
  }
  return { error: errors.join('\n\n') }
}
