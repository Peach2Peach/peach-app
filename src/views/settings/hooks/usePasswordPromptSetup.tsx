import { useMemo, useState } from 'react';
import { Keyboard } from 'react-native'
import shallow from 'zustand/shallow'

import { HelpIcon } from '../../../components/icons'
import { useHeaderSetup, useNavigation, useValidatedState } from '../../../hooks'
import { useShowHelp } from '../../../hooks/useShowHelp'
import { useSettingsStore } from '../../../store/settingsStore'
import { backupAccount } from '../../../utils/account'
import i18n from '../../../utils/i18n'

const passwordRules = { required: true, password: true }

export const usePasswordPromptSetup = (onSuccess: () => void) => {
  const navigation = useNavigation()

  const [setShowBackupReminder, setLastFileBackupDate] = useSettingsStore(
    (state) => [state.setShowBackupReminder, state.setLastFileBackupDate],
    shallow,
  )
  const showPopup = useShowHelp('yourPassword')
  useHeaderSetup({
    title: i18n('settings.backups.fileBackup.title'),
    icons: [{ iconComponent: <HelpIcon />, onPress: showPopup }],
  })

  const [password, setPassword, passwordIsValid, passwordError] = useValidatedState<string>('', passwordRules)
  const [passwordRepeat, setPasswordRepeat, passwordRepeatIsValid, passwordRepeatError] = useValidatedState<string>(
    '',
    passwordRules,
  )

  const [isBackingUp, setIsBackingUp] = useState(false)

  const passwordsMatch = useMemo(() => password === passwordRepeat, [password, passwordRepeat])
  const validate = () => !!password && !!passwordRepeat && passwordIsValid && passwordsMatch

  const startAccountBackup = () => {
    if (isBackingUp || !validate()) return

    Keyboard.dismiss()

    setIsBackingUp(true)
    setLastFileBackupDate(Date.now())
    setShowBackupReminder(false)
    backupAccount({
      password,
      onSuccess: () => {
        setIsBackingUp(false)
        onSuccess()
        navigation.navigate('backupCreated')
      },
      onCancel: () => {
        setIsBackingUp(false)
      },
      onError: () => {
        setIsBackingUp(false)
      },
    })
  }
  return {
    setPassword,
    password,
    passwordIsValid,
    passwordError,
    passwordRepeat,
    passwordRepeatIsValid,
    setPasswordRepeat,
    passwordRepeatError,
    validate,
    startAccountBackup,
  }
}
