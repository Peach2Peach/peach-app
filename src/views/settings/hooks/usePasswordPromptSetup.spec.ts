import { act, renderHook } from '@testing-library/react-native'
import { Keyboard } from 'react-native'
import { usePasswordPromptSetup } from './usePasswordPromptSetup'
import { NavigationWrapper, navigateMock } from '../../../../tests/unit/helpers/NavigationWrapper'
import { useHeaderState } from '../../../components/header/store'

const setShowBackupReminderMock = jest.fn()
const setLastFileBackupDateMock = jest.fn()
const useSettingsStoreMock = jest.fn((selector, _compareFn) =>
  selector({ setShowBackupReminder: setShowBackupReminderMock, setLastFileBackupDate: setLastFileBackupDateMock }),
)
jest.mock('../../../store/settingsStore', () => ({
  useSettingsStore: (selector: any, compareFn: any) => useSettingsStoreMock(selector, compareFn),
}))

const backupAccountMock = jest.fn()
jest.mock('../../../utils/account', () => ({
  backupAccount: (...args: any) => backupAccountMock(...args),
}))
const onSuccessMock = jest.fn()

const password = 'password'
const passwordRepeat = 'password'
describe('usePasswordPromptSetup', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  it('returns default correct values', () => {
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    expect(result.current.setPassword).toBeInstanceOf(Function)
    expect(result.current.password).toBe('')
    expect(result.current.passwordIsValid).toBeFalsy()
    expect(result.current.passwordError).toHaveLength(2)
    expect(result.current.setPasswordRepeat).toBeInstanceOf(Function)
    expect(result.current.passwordRepeat).toBe('')
    expect(result.current.passwordRepeatIsValid).toBeFalsy()
    expect(result.current.passwordRepeatError).toHaveLength(2)
    expect(result.current.validate).toBeInstanceOf(Function)
    expect(result.current.startAccountBackup).toBeInstanceOf(Function)
  })
  it('should set up header correctly', () => {
    renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    expect(useHeaderState.getState().title).toBe('backups')
    expect(useHeaderState.getState().icons?.[0].id).toBe('helpCircle')
  })
  it('should set the password', () => {
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPassword(password)
    })
    expect(result.current.password).toBe(password)
  })

  it('should set the password repeat', () => {
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPasswordRepeat(passwordRepeat)
    })
    expect(result.current.passwordRepeat).toBe(passwordRepeat)
  })

  it('should validate the passwords', () => {
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPassword(password)
      result.current.setPasswordRepeat(passwordRepeat)
    })

    expect(result.current.validate()).toBeTruthy()
  })
})

// eslint-disable-next-line max-lines-per-function
describe('startAccountBackup', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  it('should not start backup when passwords are not valid', () => {
    const keyboardSpy = jest.spyOn(Keyboard, 'dismiss')
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.startAccountBackup()
    })

    expect(keyboardSpy).not.toHaveBeenCalled()
  })
  it('should dismiss keyboard when passwords are valid', () => {
    jest.clearAllMocks()
    const keyboardSpy = jest.spyOn(Keyboard, 'dismiss')
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPassword(password)
      result.current.setPasswordRepeat(passwordRepeat)
    })
    act(() => {
      result.current.startAccountBackup()
    })

    expect(keyboardSpy).toHaveBeenCalled()
  })

  it('should set the last file backup date to now', () => {
    const now = new Date('2021-01-01')
    jest.spyOn(Date, 'now').mockImplementationOnce(() => now.getTime())
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPassword(password)
      result.current.setPasswordRepeat(passwordRepeat)
    })
    act(() => {
      result.current.startAccountBackup()
    })

    expect(setLastFileBackupDateMock).toHaveBeenCalledWith(now.getTime())
  })

  it('should set the show backup reminder to false', () => {
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPassword(password)
      result.current.setPasswordRepeat(passwordRepeat)
    })
    act(() => {
      result.current.startAccountBackup()
    })

    expect(setShowBackupReminderMock).toHaveBeenCalledWith(false)
  })

  it('should call backupAccount with the password', () => {
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPassword(password)
      result.current.setPasswordRepeat(passwordRepeat)
    })
    act(() => {
      result.current.startAccountBackup()
    })

    expect(backupAccountMock).toHaveBeenCalled()
    expect(backupAccountMock).toHaveBeenCalledWith(expect.objectContaining({ password }))
  })

  it('should handle a successful backup', () => {
    backupAccountMock.mockImplementationOnce(({ onSuccess }) => onSuccess())
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPassword(password)
      result.current.setPasswordRepeat(passwordRepeat)
    })
    act(() => {
      result.current.startAccountBackup()
    })

    expect(onSuccessMock).toHaveBeenCalled()
    expect(navigateMock).toHaveBeenCalledWith('backupCreated')
  })

  it('should handle a failed backup', () => {
    backupAccountMock.mockImplementationOnce(({ onError }) => onError())
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPassword(password)
      result.current.setPasswordRepeat(passwordRepeat)
    })
    act(() => {
      result.current.startAccountBackup()
    })

    expect(onSuccessMock).not.toHaveBeenCalled()
    expect(navigateMock).not.toHaveBeenCalledWith('backupCreated')
  })

  it('should handle a cancelled backup', () => {
    backupAccountMock.mockImplementationOnce(({ onCancel }) => onCancel())
    const { result } = renderHook(usePasswordPromptSetup, { wrapper: NavigationWrapper, initialProps: onSuccessMock })
    act(() => {
      result.current.setPassword(password)
      result.current.setPasswordRepeat(passwordRepeat)
    })
    act(() => {
      result.current.startAccountBackup()
    })

    expect(onSuccessMock).not.toHaveBeenCalled()
    expect(navigateMock).not.toHaveBeenCalledWith('backupCreated')
  })
})
