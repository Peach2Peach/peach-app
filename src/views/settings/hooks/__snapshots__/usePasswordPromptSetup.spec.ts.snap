// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePasswordPromptSetup should set up header correctly 1`] = `
<OldHeader
  icons={
    [
      {
        "color": "#099DE2",
        "id": "helpCircle",
        "onPress": [Function],
      },
    ]
  }
  title="backups"
/>
`;
